### ISSUE TITLE
The fullscreen flag on BrowserWindow doesn’t work on Windows

### ISSUE THREAD

**Issue by bvahdat** [atom/electron#2047]
However the same codebase on OS-X with the set `fullscreen: true` option works as expected.

**Comment by frankhale** [comment]
This one is interesting. When the option is set on the browser window the following function is called. Because the window is not yet visible the second part of the IsVisible check is called. The follow API call does not appear to work on Windows.

```cpp
window_->native_widget_private()->ShowWithWindowState(ui::SHOW_STATE_FULLSCREEN);
```

Full function:

```cpp
void NativeWindowViews::SetFullScreen(bool fullscreen) {
  if (IsVisible())
    window_->SetFullscreen(fullscreen);
  else
    window_->native_widget_private()->ShowWithWindowState(
        ui::SHOW_STATE_FULLSCREEN);
#if defined(OS_WIN)
  // There is no native fullscreen state on Windows.
  if (fullscreen)
    NotifyWindowEnterFullScreen();
  else
    NotifyWindowLeaveFullScreen();
#endif
}
```

I found that if I commented out the IsVisible check and just allowed it to set fullscreen then it worked as expected.

```cpp
void NativeWindowViews::SetFullScreen(bool fullscreen) {
  //if (IsVisible())
  window_->SetFullscreen(fullscreen);
  //else
  //  window_->native_widget_private()->ShowWithWindowState(
  //      ui::SHOW_STATE_FULLSCREEN);
#if defined(OS_WIN)
  // There is no native fullscreen state on Windows.
  if (fullscreen)
    NotifyWindowEnterFullScreen();
  else
    NotifyWindowLeaveFullScreen();
#endif
}
```

So that said I guess our princess is in the following castle:

```cpp
// Do we need to call this? If so, need to figure out why it's currently broken.
window_->native_widget_private()->ShowWithWindowState(ui::SHOW_STATE_FULLSCREEN);
```

**Comment by frankhale** [comment]
Additionally if I just change ui::SHOW_STATE_FULLSCREEN to ui::SHOW_STATE_MAXIMIZED the call to ShowWithWindowState succeeds and we get a corresponding result. I think the problem is that ui::SHOW_STATE_FULLSCREEN is not working for some reason.

**Comment by ankitaggarwal011** [comment]
@frankhale, In the SetFullScreen method, there is another piece of code which is executed in case of the windows operating system with a comment that "There is no native fullscreen state on Windows.". What is the significance of that code? Could this bug/error be related to that section?

**Comment by frankhale** [comment]
This appears to be triggering events so that you can be notified when fullscreen was set or left.

```cpp
#if defined(OS_WIN)
  // There is no native fullscreen state on Windows.
  if (fullscreen)
    NotifyWindowEnterFullScreen();
  else
    NotifyWindowLeaveFullScreen();
#endif
```

**Comment by ankitaggarwal011** [comment]
@frankhale , Yeah, so the problem is with SHOW_STATE_FULLSCREEN. The SHOW_STATE_FULLSCREEN state has not been supported in Windows.

This could be very helpful, [](https://codereview.chromium.org/1138383008)

**Comment by frankhale** [comment]
:+1:  Nice find!

**Comment by ankitaggarwal011** [comment]
@frankhale Thanks, Are you onto solving this bug? I quickly browsed through the code, but I am unable to locate some of the header files and find the SHOW_STATE_FULLSCREEN declaration. Could you please help here?

**Comment by frankhale** [comment]
I believe we can just elminate the call which sets SHOW_STATE_FULLSCREEN. If we comment out the following lines in this function then everything works as expected from my initial tests.

```cpp
void NativeWindowViews::SetFullScreen(bool fullscreen) {
  //if (IsVisible())
  window_->SetFullscreen(fullscreen);
  //else
  //  window_->native_widget_private()->ShowWithWindowState(
  //      ui::SHOW_STATE_FULLSCREEN);
#if defined(OS_WIN)
  // There is no native fullscreen state on Windows.
  if (fullscreen)
    NotifyWindowEnterFullScreen();
  else
    NotifyWindowLeaveFullScreen();
#endif
}
```

**Comment by frankhale** [comment]
However, the reason I did not submit a PR is because I don't know if this is desired. I'd like to know why this is being called:

```cpp
window_->native_widget_private()->ShowWithWindowState(ui::SHOW_STATE_FULLSCREEN); 
```

**Comment by ankitaggarwal011** [comment]
If we recap this and evaluate the information we currently have, IsVisible detects whether the browser window is visible and chooses this option when the window is not visible (it also ultimately results in full screen on other OS expect Windows):

```c++
window_->native_widget_private()->ShowWithWindowState(ui::SHOW_STATE_FULLSCREEN); 
```

Now, this seems to work perfectly well on other operating systems expect Windows. But we found out here (https://codereview.chromium.org/1138383008) why it's not working with Windows. It is because the SHOW_STATE_FULLSCREEN state has not been supported in Windows. This is also validated since you tried commenting this line of code and it worked well on Windows.

So, now in order to eliminate this bug, we can simply comment the lines of code as you mentioned and get the required behavior. However, have you tested the code after commenting out the mentioned lines with other operating systems. It could be required with other OS and commenting those lines might break their functionality. So, I think a better solution would be check whether the OS is Windows and perform the task accordingly like here ():

```c++
void NativeWindowViews::SetFullScreen(bool fullscreen) {
#if defined(OS_WIN)
  // There is no native fullscreen state on Windows.
  window_->SetFullscreen(fullscreen);
  if (fullscreen)
    NotifyWindowEnterFullScreen();
  else
    NotifyWindowLeaveFullScreen();
#else
  if (IsVisible())
    window_->SetFullscreen(fullscreen);
  else
    window_->native_widget_private()->ShowWithWindowState(
        ui::SHOW_STATE_FULLSCREEN);
#endif
}
```
Maybe someone else from core team can shed some more light. Is there anyone who could confirm this? @frankhale

**Comment by frankhale** [comment]
I think your change is more inclusive and would represent a nice compromise at least until the SHOW_STATE_FULLSCREEN for Windows lands in Chrome. 

However simply calling window_->SetFullscreen without the IsVisible() check seems to do the trick but I cannot confirm this behavior on Linux or Mac. 

@zcbenz, what are your thoughts on this?

**Comment by ankitaggarwal011** [comment]
I have added this change in my forked version of repository on a separate branch (https://github.com/ankitaggarwal011/electron/tree/windows-fullscreen-bug) and built it from source on my Windows 7 machine. Everything seems to be working fine. Should I make a PR on this?

I also have a question, once the window goes into fullscreen, there doesn't seem to be an exit fullscreen method other than closing/exiting the window. Is this another bug or this is by design?

@frankhale @zcbenz

**Comment by frankhale** [comment]
Sure make a PR for it =)

**Comment by ankitaggarwal011** [comment]
I was just waiting for comments from project maintainers before making PR. But, I would go ahead and make a PR. Thanks @frankhale .

Also, what do you think about this? Once the window goes into fullscreen, there doesn't seem to be an exit fullscreen method other than closing/exiting the window. Is this another bug or this is by design?

**Comment by frankhale** [comment]
At least in the the case of the default app that Electron ships with the F11 key enters and exits fullscreen. I think it's left as an exercise to the developer to enter and exit fullscreen with whatever controls are desirable.

**Comment by frankhale** [comment]
I say just make the PR and wait for @zcbenz to comment on it or accept it.

**Comment by ankitaggarwal011** [comment]
I understand about your reason about providing developers the flexibility to include exit fullscreen option or not. I have created the PR. I will wait for @zcbenz to have a look on it. Thanks for your help @frankhale .

**Comment by zcbenz** [comment]
Fixed by #2074.
